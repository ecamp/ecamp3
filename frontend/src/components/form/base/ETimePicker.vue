<!--
Displays a field as a time picker (can be used with v-model)
Allows 15min steps only
-->

<template>
  <base-picker
    :icon="icon"
    :value="value"
    :format="format"
    :format-picker="formatPicker"
    :parse="parse"
    :parse-picker="parsePicker"
    v-bind="$attrs"
    :vee-id="veeId"
    :vee-rules="veeRules"
    @input="$emit('input', $event)"
  >
    <template slot-scope="picker">
      <v-time-picker
        :value="picker.value || ''"
        :allowed-minutes="allowedStep"
        format="24hr"
        scrollable
        @input="picker.on.input"
      >
        <v-spacer />
        <v-btn text color="primary" data-testid="action-cancel" @click="picker.on.close">
          {{ $tc('global.button.cancel') }}
        </v-btn>
        <v-btn text color="primary" data-testid="action-ok" @click="picker.on.save">
          {{ $tc('global.button.ok') }}
        </v-btn>
      </v-time-picker>
    </template>

    <!-- passing the append slot through -->
    <template #append>
      <slot name="append" />
    </template>
  </base-picker>
</template>

<script>
import BasePicker from './BasePicker.vue'
import { HTML5_FMT } from '@/common/helpers/dateFormat.js'
import { formComponentMixin } from '@/mixins/formComponentMixin.js'

export default {
  name: 'ETimePicker',
  components: { BasePicker },
  mixins: [formComponentMixin],
  props: {
    icon: { type: String, required: false, default: 'mdi-clock-outline' },
    value: { type: [Number, String], required: true },

    // format in which the `value` property is being provided & input events are triggered
    valueFormat: { type: [String, Array], default: 'YYYY-MM-DDTHH:mm:ssZ' },
  },
  methods: {
    allowedStep: (m) => m % 15 === 0,

    /**
     * override time on value but keep date
     */
    setTimeOnValue(time) {
      // current value as DayJS
      let valueDateTime = this.getValueAsDateTime(this.value)

      // override time
      if (valueDateTime && valueDateTime.isValid()) {
        valueDateTime = valueDateTime
          .hour(time.hour())
          .minute(time.minute())
          .second(time.second())
          .millisecond(time.millisecond())
      } else {
        valueDateTime = time
      }

      // return in value format
      return valueDateTime.format(this.valueFormat)
    },

    /**
     * returns val as DayJS object
     */
    getValueAsDateTime(val) {
      return this.$date.utc(val, this.valueFormat)
    },

    /**
     * Format internal value for display in the UI
     */
    format(val) {
      if (val !== '') {
        return this.getValueAsDateTime(val).format('LT')
      }
      return ''
    },

    /**
     * Format internal value for the popup component
     */
    formatPicker(val) {
      if (val !== '') {
        return this.getValueAsDateTime(val).format(HTML5_FMT.TIME)
      }
      return ''
    },

    /**
     * Parse a user-supplied value into the internal format
     */
    parse(val) {
      if (val) {
        const parsedDateTime = this.$date.utc(val, 'LT')
        if (parsedDateTime.isValid() && parsedDateTime.format('LT') === val) {
          const newValue = this.setTimeOnValue(parsedDateTime)
          return Promise.resolve(newValue)
        } else {
          return Promise.reject(new Error('invalid format'))
        }
      } else {
        return Promise.resolve('')
      }
    },

    /**
     * Parse the value from the popup component into the internal format
     */
    parsePicker(val) {
      if (val) {
        const parsedDateTime = this.$date.utc(val, HTML5_FMT.TIME)
        if (parsedDateTime.isValid() && parsedDateTime.format(HTML5_FMT.TIME) === val) {
          const newValue = this.setTimeOnValue(parsedDateTime)
          return Promise.resolve(newValue)
        } else {
          return Promise.reject(new Error('invalid format'))
        }
      } else {
        return Promise.resolve('')
      }
    },
  },
}
</script>

<style scoped></style>
