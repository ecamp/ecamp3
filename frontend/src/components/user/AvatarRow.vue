<template>
  <div class="avatarrow" :style="avatarrow">
    <div v-for="user in users" :key="user.id" class="avataritem">
      <UserAvatar :size="Number(size)" :user="user" />
    </div>
  </div>
</template>

<script>
export default {
  name: 'AvatarRow',
  props: {
    users: { type: Array, default: () => [] },
    size: { type: [Number, String], default: 20 },
  },
  computed: {
    maxWidth() {
      return (this.users?.length - 1) * (Number(this.size) * 0.25) + Number(this.size)
    },
    avatarrow() {
      return {
        'max-width': `${this.maxWidth}px`,
        'font-size': `${this.size}px`,
      }
    },
  },
}
</script>

<style scoped lang="scss">
.avatarrow {
  display: flex;
  flex-direction: row-reverse;
  gap: 0.75em;
  padding-left: 0.5em;
  padding-right: 0.5em;
  transition: gap 0.25s ease;
}

@media #{map-get($display-breakpoints, 'md-and-up')} {
  .avatarrow {
    gap: 1.1em;
  }
}

.avatarrow:hover {
  gap: 1.1em;
}

.avataritem {
  display: grid;
  width: 0;
  place-content: center;
  text-decoration: none;
}
</style>
