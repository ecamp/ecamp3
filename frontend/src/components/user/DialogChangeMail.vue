<template>
  <dialog-form
    v-model="showDialog"
    :title="$tc('components.user.dialogChangeMail.title')"
    :submit-action="status === 'initial' ? sendChangeMailRequest : null"
    :cancel-action="close"
    cancel-label="global.button.close"
    :cancel-visible="status !== 'initial'"
    submit-color="success"
  >
    <template #activator="scope">
      <slot name="activator" v-bind="scope" />
    </template>
    <div v-if="status === 'initial'">
      <e-text-field
        v-model="entityData.newEmail"
        :name="$tc('entity.user.fields.email')"
        vee-rules="email|required"
        append-icon="mdi-at"
        autofocus
      />
      <p class="mt-5">
        {{ $tc('components.user.dialogChangeMail.message') }}
      </p>
    </div>
    <div v-if="status == 'success'">
      {{ $tc('components.user.dialogChangeMail.success') }}
    </div>
    <div v-if="status == 'error'">
      {{ $tc('components.user.dialogChangeMail.error') }}
    </div>
  </dialog-form>
</template>

<script>
import DialogForm from '@/components/dialog/DialogForm.vue'
import DialogBase from '@/components/dialog/DialogBase.vue'

export default {
  name: 'DialogChangeMail',
  components: { DialogForm },
  extends: DialogBase,
  data() {
    return {
      status: '',
      entityProperties: ['newEmail'],
    }
  },
  watch: {
    // copy data whenever dialog is opened
    showDialog: function (showDialog) {
      if (showDialog) {
        this.status = 'initial'
        this.loadEntityData(this.$store.state.auth.user.profile()._meta.self)
      }
    },
  },
  methods: {
    async sendChangeMailRequest() {
      await this.api
        .patch(this.entityUri, this.entityData)
        .then(() => {
          this.status = 'success'
        })
        .catch(() => {
          this.status = 'error'
        })
    },
  },
}
</script>
