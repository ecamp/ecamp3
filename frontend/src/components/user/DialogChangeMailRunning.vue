<template>
  <dialog-form
    v-model="showDialog"
    :title="$tc('components.user.dialogChangeMailRunning.title')"
    :cancel-action="status === 'initial' ? null : close"
    cancel-label="global.button.close"
  >
    <div v-if="status === 'initial'">
      <v-progress-circular indeterminate />
      {{ $tc('components.user.dialogChangeMailRunning.message') }}
    </div>
    <div v-if="status === 'success'">
      <v-icon>mdi mdi-check</v-icon>
      {{ $tc('components.user.dialogChangeMailRunning.success') }}
    </div>
    <div v-if="status === 'error'">
      {{ $tc('components.user.dialogChangeMailRunning.error') }}
    </div>
  </dialog-form>
</template>

<script>
import DialogForm from '@/components/dialog/DialogForm.vue'
import DialogBase from '@/components/dialog/DialogBase.vue'

export default {
  name: 'DialogChangeMailRunning',
  components: { DialogForm },
  extends: DialogBase,
  props: {
    emailVerificationKey: { type: String, required: false, default: null },
  },
  data() {
    return {
      status: '',
    }
  },
  watch: {
    showDialog: function (showDialog) {
      if (!showDialog) {
        this.$router.push({ name: 'profile' })
      }
    },
  },
  mounted() {
    if (this.emailVerificationKey) {
      this.status = 'initial'
      this.showDialog = true

      this.$auth
        .user()
        .profile()
        ._meta.load.then((p) => {
          p.$patch({
            untrustedEmailKey: this.emailVerificationKey,
          })
            .then(() => {
              this.status = 'success'
            })
            .catch(() => {
              this.status = 'error'
            })
        })
    }
  },
}
</script>
