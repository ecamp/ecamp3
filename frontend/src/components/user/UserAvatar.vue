<template>
  <v-avatar :size="size" :color="color">
    <span class="white--text font-weight-medium" :style="style">{{ initials }}</span>
  </v-avatar>
</template>
<script>
import campCollaborationInitials from '@/common/helpers/campCollaborationInitials.js'
import campCollaborationColor from '@/common/helpers/campCollaborationColor.js'
import userColor from '@/common/helpers/userColor.js'
import userInitials from '@/common/helpers/userInitials.js'

export default {
  name: 'UserAvatar',
  props: {
    size: { type: Number, required: false, default: 48 },
    user: { type: Object, default: null },
    campCollaboration: { type: Object, default: null },
  },
  computed: {
    isLoading() {
      return (this.user || this.campCollaboration)?._meta.loading
    },
    color() {
      if (this.isLoading) {
        return 'rgba(0,0,0,0)'
      }
      return this.user
        ? userColor(this.user)
        : campCollaborationColor(this.campCollaboration)
    },
    initials() {
      if (this.isLoading) {
        return ''
      }
      return this.user
        ? userInitials(this.user)
        : campCollaborationInitials(this.campCollaboration)
    },
    style() {
      return {
        fontSize: this.size / 2.5 + 'px',
        fontWeight: 400,
        letterSpacing: '1px',
        marginRight: '-1.5px',
      }
    },
  },
}
</script>
