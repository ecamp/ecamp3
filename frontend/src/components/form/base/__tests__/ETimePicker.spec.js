import { screen, waitFor } from '@testing-library/vue'
import { render, setTestLocale, snapshotOf } from '@/test/renderWithVuetify.js'
import user from '@testing-library/user-event'
import ETimePicker from '../ETimePicker.vue'

describe('An ETimePicker', () => {
  const TIME1_ISO = '2037-07-18T09:52:00+00:00'
  const TIME1_HHMM = '09:52'
  const TIME2_ISO = '2037-07-18T18:33:00+00:00'
  const TIME3_ISO = '2037-07-18T00:52:00+00:00'

  const localeData = {
    de: {
      time1: '09:52',
      time2: '18:33',
      time3: '00:52',
      firstHour: '0',
      labelText: 'Dialog öffnen um eine Zeit für test zu wählen',
      closeButton: 'Schliessen',
      timeInWrongLocale: '9:52 AM',
      validationMessage: 'Ungültiges Format, bitte gib die Zeit im Format HH:MM ein',
    },
    en: {
      time1: '9:52 AM',
      time2: '6:33 PM',
      time3: '12:52 AM',
      firstHour: '12',
      labelText: 'Open dialog to select a time for test',
      closeButton: 'Close',
      timeInWrongLocale: '09:52',
      validationMessage:
        'Invalid format, please enter the time in the format HH:MM AM/PM',
    },
  }

  describe.each(Object.entries(localeData))('in locale %s', (locale, data) => {
    beforeEach(() => {
      setTestLocale(locale)
    })

    it('renders the component', async () => {
      // given

      // when
      render(ETimePicker, {
        props: {
          value: TIME1_ISO,
          name: 'test',
        },
      })

      // then
      await screen.findByDisplayValue(data.time1)
      screen.getByLabelText(data.labelText)
    })

    it('looks like a time picker', async () => {
      // given

      // when
      const { container } = render(ETimePicker, {
        props: { value: TIME1_ISO, name: 'test' },
      })

      // then
      expect(snapshotOf(container)).toMatchSnapshot('pickerclosed')

      // when
      await user.click(screen.getByLabelText(data.labelText))

      // then
      await screen.findByText(data.closeButton)
      expect(snapshotOf(container)).toMatchSnapshot('pickeropen')
    })

    it('allows setting a different valueFormat', async () => {
      // given

      // when
      render(ETimePicker, {
        props: {
          value: TIME1_HHMM,
          name: 'test',
          valueFormat: 'HH:mm',
        },
      })

      // then
      await screen.findByDisplayValue(data.time1)
      screen.getByLabelText(data.labelText)
    })

    it('does not open the picker when the text field is clicked', async () => {
      // given
      render(ETimePicker, {
        props: { value: TIME1_ISO, name: 'test' },
      })
      const inputField = await screen.findByDisplayValue(data.time1)

      // when
      await user.click(inputField)

      // then
      // menu should not appear
      await expect(screen.findByText(data.closeButton)).rejects.toThrow(
        /Unable to find an element with the text/
      )
    })

    it('opens the picker when the icon button is clicked', async () => {
      // given
      render(ETimePicker, {
        props: { value: TIME1_ISO, name: 'test' },
      })
      const button = await screen.getByLabelText(data.labelText)

      // when
      await user.click(button)

      // then
      await waitFor(async () => {
        expect(await screen.findByText(data.closeButton)).toBeVisible()
      })
    })

    it('closes the picker when clicking the close button', async () => {
      // given
      render(ETimePicker, {
        props: { value: TIME1_ISO, name: 'test' },
      })
      const button = await screen.getByLabelText(data.labelText)
      await user.click(button)
      await waitFor(async () => {
        expect(await screen.findByText(data.closeButton)).toBeVisible()
      })
      const closeButton = screen.getByText(data.closeButton)

      // when
      await user.click(closeButton)

      // then
      await waitFor(async () => {
        expect(await screen.queryByText(data.closeButton)).not.toBeVisible()
      })
    })

    it('closes the picker when clicking outside', async () => {
      // given
      render(ETimePicker, {
        props: { value: TIME1_ISO, name: 'test' },
      })
      const button = await screen.getByLabelText(data.labelText)
      await user.click(button)
      await waitFor(async () => {
        expect(await screen.findByText(data.closeButton)).toBeVisible()
      })

      // when
      await user.click(document.body)

      // then
      await waitFor(async () => {
        expect(await screen.queryByText(data.closeButton)).not.toBeVisible()
      })
    })

    it('closes the picker when pressing escape', async () => {
      // given
      render(ETimePicker, {
        props: { value: TIME1_ISO, name: 'test' },
      })
      const button = await screen.getByLabelText(data.labelText)
      await user.click(button)
      await waitFor(async () => {
        expect(await screen.findByText(data.closeButton)).toBeVisible()
      })

      // when
      await user.click(document.body)

      // then
      await waitFor(async () => {
        expect(await screen.queryByText(data.closeButton)).not.toBeVisible()
      })
    })

    it('does not close the picker when selecting a time', async () => {
      // given
      render(ETimePicker, {
        props: { value: TIME1_ISO, name: 'test' },
      })
      const button = await screen.getByLabelText(data.labelText)
      await user.click(button)
      await waitFor(async () => {
        expect(await screen.findByText(data.closeButton)).toBeVisible()
      })

      // when
      // Click the 0th hour
      await user.click(await screen.findByText(data.firstHour))
      // Click the 45th minute
      await user.click(await screen.findByText('45'))

      // then
      // close button should stay visible
      await expect(
        waitFor(() => {
          expect(screen.queryByText(data.closeButton)).not.toBeVisible()
        })
      ).rejects.toThrow(/Received element is visible/)
    })

    describe('updates v-model when the input field is changed', async () => {
      const timeConfig1 = {
        iso: TIME1_ISO,
        localizedTime: data.time1,
        textInput: data.time1,
      }
      const timeConfig2 = {
        iso: TIME2_ISO,
        localizedTime: data.time2,
        textInput: data.time2,
      }
      const timeConfig3 = {
        iso: TIME3_ISO,
        localizedTime: data.time3,
        textInput: data.time3,
      }
      it.each([
        {
          from: timeConfig1,
          to: timeConfig2,
        },
        {
          from: timeConfig2,
          to: timeConfig1,
        },
        {
          from: timeConfig1,
          to: timeConfig3,
        },
        //with leading zero
        {
          from: timeConfig1,
          to: {
            ...timeConfig2,
            textInput: '0' + timeConfig2.localizedTime,
          },
        },
        {
          from: timeConfig1,
          to: {
            ...timeConfig2,
            textInput: '0000' + timeConfig2.localizedTime,
          },
        },
        {
          from: timeConfig2,
          to: {
            ...timeConfig1,
            textInput: '0' + timeConfig1.localizedTime,
          },
        },
        {
          from: timeConfig2,
          to: {
            ...timeConfig1,
            textInput: '00000' + timeConfig1.localizedTime,
          },
        },
        {
          from: timeConfig1,
          to: {
            ...timeConfig3,
            textInput: '0' + timeConfig3.localizedTime,
          },
        },
        {
          from: timeConfig1,
          to: {
            ...timeConfig3,
            textInput: '00000' + timeConfig3.localizedTime,
          },
        },
      ])(
        `from $from.localizedTime to $to.textInput`,
        async ({
          from: { iso: fromIso, localizedTime: fromLocalizedTime },
          to: { iso: toIso, localizedTime: toLocalizedTime, textInput },
        }) => {
          // given
          const { emitted } = render(ETimePicker, {
            props: { value: fromIso, name: 'test' },
          })
          const inputField = await screen.findByDisplayValue(fromLocalizedTime)

          // when
          await user.clear(inputField)
          await user.keyboard(textInput)

          // then
          await waitFor(async () => {
            const events = emitted().input
            // some input events were fired
            expect(events.length).toBeGreaterThan(0)
            // the last one included the parsed version of our entered time
            expect(events[events.length - 1]).toEqual([toIso])
          })
          // Our entered time should be visible...
          screen.getByDisplayValue(toLocalizedTime)
          // ...and stay visible
          await expect(
            waitFor(() => {
              expect(screen.getByDisplayValue(toLocalizedTime)).not.toBeVisible()
            })
          ).rejects.toThrow(/Received element is visible/)
        }
      )
    })

    it('updates v-model when a new time is selected in the picker', async () => {
      // given
      const { emitted } = render(ETimePicker, {
        props: { value: TIME1_ISO, name: 'test' },
      })
      await screen.findByDisplayValue(data.time1)
      const button = await screen.getByLabelText(data.labelText)

      // when
      // click the button to open the picker
      await user.click(button)
      // Click the 0th hour. We can only click this one, because
      // testing library is missing the vuetify styles, and all the
      // number elements overlap
      await user.click(await screen.findByText(data.firstHour))
      // click the close button
      await user.click(screen.getByText(data.closeButton))

      // then
      await waitFor(() => {
        expect(screen.queryByText(data.closeButton)).not.toBeVisible()
      })
      await waitFor(async () => {
        const events = emitted().input
        // some input events were fired
        expect(events.length).toBeGreaterThan(0)
        // the last one included the parsed version of our entered time
        expect(events[events.length - 1]).toEqual([TIME3_ISO])
      })
      // Our entered time should be visible...
      screen.getByDisplayValue(data.time3)
      // ...and stay visible
      await expect(
        waitFor(() => {
          expect(screen.getByDisplayValue(data.time3)).not.toBeVisible()
        })
      ).rejects.toThrow(/Received element is visible/)
    })

    describe('validates the input', async () => {
      it.each([
        data.timeInWrongLocale,
        'a' + data.time1,
        data.time2 + 'a',
        '0000:a' + data.time3,
      ])('%s', async (textInput) => {
        // given
        render(ETimePicker, {
          props: { value: TIME1_ISO, name: 'test' },
        })
        const inputField = await screen.findByDisplayValue(data.time1)

        // when
        await user.clear(inputField)
        await user.keyboard(textInput)

        // then
        await screen.findByText(data.validationMessage)
      })
    })

    it('works with invalid initialization', async () => {
      // given

      // when
      render(ETimePicker, {
        props: {
          value: 'abc',
          name: 'test',
        },
      })

      // then
      const inputField = await screen.findByDisplayValue('Invalid Date')

      // when
      await user.clear(inputField)
      await user.keyboard(data.time2)

      // then
      waitFor(() => {
        expect(screen.queryByText('Invalid Date')).not.toBeInTheDocument()
      })
    })
  })
})
