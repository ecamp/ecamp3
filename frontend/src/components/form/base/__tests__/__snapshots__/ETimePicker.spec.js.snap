// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`An ETimePicker in locale de looks like a time picker: pickerclosed 1`] = `
<div data-app>
     <div class="e-form-container"
          vee-rules>
          <div class="e-form-container"
               value="09:52">
               <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-text-field--filled v-text-field--is-booted v-text-field--enclosed">
                    <div class="v-input__prepend-outer">
                         <button class="v-icon notranslate v-icon--link mdi mdi-clock-outline theme--light"
                                 type="button"></button>
                    </div>
                    <div class="v-input__control">
                         <div class="v-input__slot">
                              <div class="v-text-field__slot">
                                   <input id
                                          type="text">
                              </div>
                         </div>
                    </div>
               </div>
          </div>
          <div class="v-menu">
               <!---->
          </div>
     </div>
</div>
`;

exports[`An ETimePicker in locale de looks like a time picker: pickeropen 1`] = `
<div data-app>
     <div class="e-form-container"
          vee-rules>
          <div class="e-form-container"
               value="09:52">
               <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-text-field--filled v-text-field--is-booted v-text-field--enclosed">
                    <div class="v-input__prepend-outer">
                         <button class="v-icon notranslate v-icon--link mdi mdi-clock-outline theme--light"
                                 type="button"></button>
                    </div>
                    <div class="v-input__control">
                         <div class="v-input__slot">
                              <div class="v-text-field__slot">
                                   <input id
                                          type="text">
                              </div>
                         </div>
                    </div>
               </div>
          </div>
          <div class="v-menu">
               <transition-stub name="scale-transition"></transition-stub>
          </div>
     </div>
     <div class="v-menu__content theme--light menuable__content__active "
          role="menu"
          style="max-height: auto; min-width: 290px; max-width: 290px; top: 12px; left: 0px; transform-origin: top left; z-index: 8; display: none;">
          <div class="v-picker v-card v-picker--time theme--light">
               <div class="v-picker__title primary">
                    <div class="v-time-picker-title">
                         <div class="v-time-picker-title__time">
                              <div class="v-picker__title__btn v-picker__title__btn--active">09</div>
                              <span>:</span>
                              <div class="v-picker__title__btn">52</div>
                         </div>
                    </div>
               </div>
               <div class="v-picker__body theme--light"
                    style="width: 290px;">
                    <transition-stub name="fade-transition">
                         <div class="v-time-picker-clock__container">
                              <div class="v-time-picker-clock theme--light">
                                   <div class="v-time-picker-clock__inner">
                                        <div class="v-time-picker-clock__hand accent"
                                             style="transform: rotate(270deg) scaleY(1);"></div>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 50%; top: 0%;">
                                             <span>0</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 75%; top: 6.698729810778062%;">
                                             <span>1</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 93.30127018922192%; top: 24.999999999999993%;">
                                             <span>2</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 100%; top: 50%;">
                                             <span>3</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 93.30127018922194%; top: 74.99999999999999%;">
                                             <span>4</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 75.00000000000001%; top: 93.30127018922192%;">
                                             <span>5</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 50.00000000000001%; top: 100%;">
                                             <span>6</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 25.000000000000014%; top: 93.30127018922194%;">
                                             <span>7</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 6.698729810778076%; top: 75.00000000000003%;">
                                             <span>8</span>
                                        </span>
                                        <span class="v-time-picker-clock__item v-time-picker-clock__item--active accent"
                                              style="left: 0%; top: 50.00000000000001%;">
                                             <span>9</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 6.6987298107780475%; top: 25.000000000000032%;">
                                             <span>10</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 24.99999999999998%; top: 6.698729810778083%;">
                                             <span>11</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 49.99999999999999%; top: 19%;">
                                             <span>12</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 65.49999999999997%; top: 23.153212482682388%;">
                                             <span>13</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 76.84678751731758%; top: 34.499999999999986%;">
                                             <span>14</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 81%; top: 49.999999999999964%;">
                                             <span>15</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 76.84678751731761%; top: 65.49999999999997%;">
                                             <span>16</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 65.50000000000004%; top: 76.84678751731758%;">
                                             <span>17</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 50.000000000000014%; top: 81%;">
                                             <span>18</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 34.50000000000003%; top: 76.84678751731761%;">
                                             <span>19</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 23.153212482682427%; top: 65.50000000000004%;">
                                             <span>20</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 19%; top: 50.000000000000014%;">
                                             <span>21</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 23.153212482682385%; top: 34.50000000000003%;">
                                             <span>22</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 34.49999999999996%; top: 23.153212482682427%;">
                                             <span>23</span>
                                        </span>
                                   </div>
                              </div>
                         </div>
                    </transition-stub>
               </div>
               <div class="v-picker__actions v-card__actions">
                    <div class="spacer"></div>
                    <button class="v-btn v-btn--text theme--light v-size--default primary--text"
                            type="button">
                         <span class="v-btn__content">
                              Abbrechen
                         </span>
                    </button>
                    <button class="v-btn v-btn--text theme--light v-size--default primary--text"
                            type="button">
                         <span class="v-btn__content">
                              In Ordnung
                         </span>
                    </button>
               </div>
          </div>
     </div>
</div>
`;

exports[`An ETimePicker in locale en looks like a time picker: pickerclosed 1`] = `
<div data-app>
     <div class="e-form-container"
          vee-rules>
          <div class="e-form-container"
               value="9:52 AM">
               <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-text-field--filled v-text-field--is-booted v-text-field--enclosed">
                    <div class="v-input__prepend-outer">
                         <button class="v-icon notranslate v-icon--link mdi mdi-clock-outline theme--light"
                                 type="button"></button>
                    </div>
                    <div class="v-input__control">
                         <div class="v-input__slot">
                              <div class="v-text-field__slot">
                                   <input id
                                          type="text">
                              </div>
                         </div>
                    </div>
               </div>
          </div>
          <div class="v-menu">
               <!---->
          </div>
     </div>
</div>
`;

exports[`An ETimePicker in locale en looks like a time picker: pickeropen 1`] = `
<div data-app>
     <div class="e-form-container"
          vee-rules>
          <div class="e-form-container"
               value="9:52 AM">
               <div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-text-field--single-line v-text-field--filled v-text-field--is-booted v-text-field--enclosed">
                    <div class="v-input__prepend-outer">
                         <button class="v-icon notranslate v-icon--link mdi mdi-clock-outline theme--light"
                                 type="button"></button>
                    </div>
                    <div class="v-input__control">
                         <div class="v-input__slot">
                              <div class="v-text-field__slot">
                                   <input id
                                          type="text">
                              </div>
                         </div>
                    </div>
               </div>
          </div>
          <div class="v-menu">
               <transition-stub name="scale-transition"></transition-stub>
          </div>
     </div>
     <div class="v-menu__content theme--light menuable__content__active "
          role="menu"
          style="max-height: auto; min-width: 290px; max-width: 290px; top: 12px; left: 0px; transform-origin: top left; z-index: 8; display: none;">
          <div class="v-picker v-card v-picker--time theme--light">
               <div class="v-picker__title primary">
                    <div class="v-time-picker-title">
                         <div class="v-time-picker-title__time">
                              <div class="v-picker__title__btn v-picker__title__btn--active">09</div>
                              <span>:</span>
                              <div class="v-picker__title__btn">52</div>
                         </div>
                    </div>
               </div>
               <div class="v-picker__body theme--light"
                    style="width: 290px;">
                    <transition-stub name="fade-transition">
                         <div class="v-time-picker-clock__container">
                              <div class="v-time-picker-clock theme--light">
                                   <div class="v-time-picker-clock__inner">
                                        <div class="v-time-picker-clock__hand accent"
                                             style="transform: rotate(270deg) scaleY(1);"></div>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 50%; top: 0%;">
                                             <span>0</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 75%; top: 6.698729810778062%;">
                                             <span>1</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 93.30127018922192%; top: 24.999999999999993%;">
                                             <span>2</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 100%; top: 50%;">
                                             <span>3</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 93.30127018922194%; top: 74.99999999999999%;">
                                             <span>4</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 75.00000000000001%; top: 93.30127018922192%;">
                                             <span>5</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 50.00000000000001%; top: 100%;">
                                             <span>6</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 25.000000000000014%; top: 93.30127018922194%;">
                                             <span>7</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 6.698729810778076%; top: 75.00000000000003%;">
                                             <span>8</span>
                                        </span>
                                        <span class="v-time-picker-clock__item v-time-picker-clock__item--active accent"
                                              style="left: 0%; top: 50.00000000000001%;">
                                             <span>9</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 6.6987298107780475%; top: 25.000000000000032%;">
                                             <span>10</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 24.99999999999998%; top: 6.698729810778083%;">
                                             <span>11</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 49.99999999999999%; top: 19%;">
                                             <span>12</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 65.49999999999997%; top: 23.153212482682388%;">
                                             <span>13</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 76.84678751731758%; top: 34.499999999999986%;">
                                             <span>14</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 81%; top: 49.999999999999964%;">
                                             <span>15</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 76.84678751731761%; top: 65.49999999999997%;">
                                             <span>16</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 65.50000000000004%; top: 76.84678751731758%;">
                                             <span>17</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 50.000000000000014%; top: 81%;">
                                             <span>18</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 34.50000000000003%; top: 76.84678751731761%;">
                                             <span>19</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 23.153212482682427%; top: 65.50000000000004%;">
                                             <span>20</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 19%; top: 50.000000000000014%;">
                                             <span>21</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 23.153212482682385%; top: 34.50000000000003%;">
                                             <span>22</span>
                                        </span>
                                        <span class="v-time-picker-clock__item"
                                              style="left: 34.49999999999996%; top: 23.153212482682427%;">
                                             <span>23</span>
                                        </span>
                                   </div>
                              </div>
                         </div>
                    </transition-stub>
               </div>
               <div class="v-picker__actions v-card__actions">
                    <div class="spacer"></div>
                    <button class="v-btn v-btn--text theme--light v-size--default primary--text"
                            type="button">
                         <span class="v-btn__content">
                              Cancel
                         </span>
                    </button>
                    <button class="v-btn v-btn--text theme--light v-size--default primary--text"
                            type="button">
                         <span class="v-btn__content">
                              Okay
                         </span>
                    </button>
               </div>
          </div>
     </div>
</div>
`;
